libs/cardano-ledger-test/src/Test/Cardano/Ledger/Generic/GenState.hs

```diff
@@ -475,66 +598,128 @@ genFreshCredential tries0 tag old = go tries0
         else pure c
 
 -- Adds to the mPoolParams and the  mPoolDistr of the Model, and the initial set of objects for Traces
-genPool :: forall era. Reflect era => GenRS era (KeyHash 'StakePool (Crypto era))
-genPool = frequencyT [(10, genNewPool), (90, pickExisting)]
+genPool :: forall era. Reflect era => GenRS era (KeyHash 'StakePool (Crypto era), PoolParams (Crypto era))
+genPool = frequencyT [(10, genNew), (90, pickExisting)]
   where
+    genNew = do
+      (kh, pp, _) <- genNewPool
+      modifyModel $ \m ->
+        m
+          { mPoolParams = Map.insert kh pp $ mPoolParams m
+          }
+      modify $ \st ->
+        st
+          { gsInitialPoolParams = Map.insert kh pp $ gsInitialPoolParams st
+          }
+      return (kh, pp)
     pickExisting = do
       _pParams <- gets (mPoolParams . gsModel)
       lift (genMapElem _pParams) >>= \case
-        Nothing -> genNewPool
-        Just poolId -> pure $ fst poolId
-    genNewPool = do
-      poolId <- genKeyHash
-      poolparam <- genPoolParams poolId
-      percent <- lift $ choose (0, 1 :: Float)
-      let stake = IndividualPoolStake @(Crypto era) (toRational percent) (_poolVrf poolparam)
-      modify
-        ( \st ->
-            st
-              { gsInitialPoolParams = Map.insert poolId poolparam (gsInitialPoolParams st),
-                gsInitialPoolDistr = Map.insert poolId stake (gsInitialPoolDistr st)
-              }
-        )
-      modifyModel
-        ( \m ->
-            m
-              { mPoolParams = Map.insert poolId poolparam (mPoolParams m),
-                mPoolDistr = Map.insert poolId stake (mPoolDistr m)
-              }
-        )
-      pure poolId
-    genPoolParams _poolId = do
-      _poolVrf <- lift arbitrary
-      _poolPledge <- lift genPositiveVal
-      _poolCost <- lift genPositiveVal
-      _poolMargin <- lift arbitrary
-      _poolRAcnt <- RewardAcnt Testnet <$> genCredential Rewrd
-      let _poolOwners = mempty
-      let _poolRelays = mempty
-      let _poolMD = SNothing
-      pure PoolParams {..}
+        Nothing -> genNew
+        Just (poolId, pp) -> pure (poolId, pp)
+
+genFreshKeyHash :: Reflect era => Int -> GenRS era (KeyHash kr (Crypto era))
+genFreshKeyHash n
+  | n <= 0 = error "Something very unlikely happened"
+  | otherwise = do
+      avoidKeys <- gets gsAvoidKey
+      kh <- genKeyHash
+      if coerceKeyRole kh `Set.member` avoidKeys
+        then genFreshKeyHash $ n - 1
+        else return kh
```

Comment:
It seems that the argument `n` is only to ensure that this function terminates, cause genKeyHash should produce uniformly distributed keys, but there is an impossibly tiny chance of a collision. If my understanding is correct it is better to bind the max number of iterations right into the function:

```suggestion
genFreshKeyHash :: Reflect era => GenRS era (KeyHash kr (Crypto era))
genFreshKeyHash = go 100 -- avoid unlikely chance of generated hash collisions.
  where
    go n
      | n <= 0 = error "Something very unlikely happened"
      | otherwise = do
          avoidKeys <- gets gsAvoidKey
          kh <- genKeyHash
          if coerceKeyRole kh `Set.member` avoidKeys
            then go $ n - 1
            else return kh
```